src/victory-util/add-events.js

```diff
@@ -14,30 +46,86 @@ export default (WrappedComponent) => {
+      const calculatedState = this.getStateChanges(nextProps, calculatedValues);
+      if (!areVictoryPropsEqual(this.calculatedState, calculatedState)) {
+        this.cacheValues(calculatedValues);
+        this.calculatedState = calculatedState;
+        return true;
+      }
+
+      // check whether props have changed
+      if (!areVictoryPropsEqual(this.props, nextProps)) {
+        this.cacheValues(calculatedValues);
+        return true;
+      }
+
+      return false;
     }
 
-    setupEvents(props) {
+    // compile all state changes from own and parent state. Order doesn't matter, as any state
+    // state change should trigger a re-render
+    getStateChanges(props, calculatedValues) {
+      const { hasEvents, getSharedEventState } = calculatedValues;
+      if (!hasEvents) { return {}; }
+
+      const getState = (key, type) => {
+        const result = defaults({}, this.getEventState(key, type), getSharedEventState(key, type));
+        return isEmpty(result) ? undefined : result;
+      };
+
+      options = options || {};
+      const components = options.components || defaultComponents;
+      return components.map((component) => {
+        if (!props.standalone && component.name === "parent") {
+           // don't check for changes on parent props for non-standalone components
+          return undefined;
+        } else {
+          return component.index !== undefined ?
+          getState(component.index, component.name) :
+          calculatedValues.dataKeys.map((key) => getState(key, component.name));
+        }
+      }).filter(Boolean);
+    }
+
+    getCalculatedValues(props) {
       const { sharedEvents } = props;
       const components = WrappedComponent.expectedComponents;
-      this.componentEvents = Events.getComponentEvents(props, components);
-      this.getSharedEventState = sharedEvents && isFunction(sharedEvents.getEventState) ?
+      const componentEvents = Events.getComponentEvents(props, components);
+      const getSharedEventState = sharedEvents && isFunction(sharedEvents.getEventState) ?
         sharedEvents.getEventState : () => undefined;
-      this.baseProps = this.getBaseProps(props);
-      this.dataKeys = Object.keys(this.baseProps).filter((key) => key !== "parent");
-      this.hasEvents = props.events || props.sharedEvents || this.componentEvents;
-      this.events = this.getAllEvents(props);
+      const baseProps = this.getBaseProps(props, getSharedEventState);
+      const dataKeys = Object.keys(baseProps).filter((key) => key !== "parent");
+      const hasEvents = props.events || props.sharedEvents || componentEvents;
+      const events = this.getAllEvents(props);
+      return {
+        componentEvents, getSharedEventState, baseProps, dataKeys, hasEvents, events
+      };
     }
 
-    getBaseProps(props) {
-      const sharedParentState = this.getSharedEventState("parent", "parent");
+    cacheValues(obj) {
+      keys(obj).forEach((key) => {
+        this[key] = obj[key];
```

Comment:
Are there any collision protections? It seems like you're adding straight to the underlying object. E.g., if someone  passes an object with:

```js
{ cacheValues: "foo" }
```

it would blow away this version method of `this.cacheValues()`